server/lib/chat/chat-models.ts

```diff
@@ -578,3 +579,94 @@ export async function findChannelByName(
     done()
   }
 }
+
+/** Returns the number of channels that match a particular condition. */
+async function getChannelsCount(
+  whereCondition?: SQLStatement,
+  withClient?: DbClient,
+): Promise<number> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = sql`
+      SELECT COUNT(*) AS count
+      FROM channels
+    `
+
+    if (whereCondition) {
+      query.append(whereCondition)
+    }
+
+    const result = await client.query<{ count: string }>(query)
+
+    return Number(result.rows[0].count)
+  } finally {
+    done()
+  }
+}
+
+/**
+ * Gets a list of chat channels for a particular user, optionally filtered by a `searchStr`. If the
+ * user is an admin, this method will return a full list of channels. Otherwise, this method will
+ * only return non-private channels and channels that the user is not joined in.
+ */
+export async function getChannelsForUser(
+  userId: SbUserId,
+  isAdmin: boolean,
+  limit: number,
+  pageNumber: number,
+  searchStr?: string,
+  withClient?: DbClient,
+): Promise<{ channels: ChannelInfo[]; total: number }> {
+  const { client, done } = await db(withClient)
+  try {
+    const query = isAdmin
+      ? sql`
+        SELECT c.*, COUNT(cu.user_id) AS user_count
+        FROM channels c
+        INNER JOIN channel_users cu ON c.id = cu.channel_id
+      `
+      : sql`
+        WITH unjoined_channels AS (
+          SELECT DISTINCT(channel_id)
+          FROM channel_users
+          WHERE channel_id NOT IN (SELECT channel_id FROM channel_users WHERE user_id = ${userId})
+        )
+        SELECT c.*, COUNT(cu.user_id) AS user_count
+        FROM channels c
+        INNER JOIN unjoined_channels ON c.id = unjoined_channels.channel_id
+        INNER JOIN channel_users cu ON c.id = cu.channel_id
+      `
+
+    let whereCondition: SQLStatement | undefined
+    if (!isAdmin) {
+      whereCondition = sql`WHERE private = false`
+    }
+    if (searchStr) {
+      const escapedStr = `%${searchStr.replace(/[_%\\]/g, '\\$&')}%`
+
+      if (whereCondition) {
+        whereCondition.append(sql` AND name ILIKE ${escapedStr}`)
+      } else {
+        whereCondition = sql`WHERE name ILIKE ${escapedStr}`
+      }
+
+      query.append(whereCondition)
+    }
+
+    query.append(sql`
+      GROUP BY c.id
+      ORDER BY user_count DESC, name
+      LIMIT ${limit}
+      OFFSET ${pageNumber * limit}
+    `)
+
+    const result = await client.query<DbChannel>(query)
+
+    const channels = await Promise.all(result.rows.map(row => convertChannelFromDb(row)))
```

Comment:
Why is this a Promise.all? This function does not return a promise (and it should absolutely never be async, so I'm at a loss as to why this was ever thought to be needed)

Reply:
Copy-paste fail :d